File: app/src/main/java/com/benny/openlauncher/widget/SearchBar.java
package com.benny.openlauncher.widget;

import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatEditText;
import android.support.v7.widget.AppCompatImageView;
import android.support.v7.widget.CardView;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.TextWatcher;
import android.text.style.RelativeSizeSpan;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.view.WindowInsets;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.benny.openlauncher.R;
import com.benny.openlauncher.interfaces.AppUpdateListener;
import com.benny.openlauncher.manager.Setup;
import com.benny.openlauncher.model.App;
import com.benny.openlauncher.model.Item;
import com.benny.openlauncher.util.AppSettings;
import com.benny.openlauncher.util.DragAction;
import com.benny.openlauncher.util.DragHandler;
import com.benny.openlauncher.util.Tool;
import com.benny.openlauncher.viewutil.CircleDrawable;
import com.benny.openlauncher.viewutil.IconLabelItem;
import com.mikepenz.fastadapter.IItemAdapter;
import com.mikepenz.fastadapter.commons.adapters.FastItemAdapter;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.threeten.bp.ZonedDateTime;
import org.threeten.bp.format.DateTimeFormatter;

import java.sql.Date;
import java.text.Normalizer;
import java.text.Normalizer.Form;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;

public class SearchBar extends FrameLayout {
    private static Logger LOG = LoggerFactory.getLogger("SearchBar");

    private static final long ANIM_TIME = 200;
    public TextView _searchClock;
    public AppCompatImageView _switchButton;
    public AppCompatImageView _searchButton;
    public AppCompatEditText _searchInput;
    public RecyclerView _searchRecycler;
    private CircleDrawable _icon;
    private CardView _searchCardContainer;
    private FastItemAdapter<IconLabelItem> _adapter = new FastItemAdapter<>();
    private CallBack _callback;
    private boolean _expanded;
    private int _searchClockTextSize = 28;
    private float _searchClockSubTextFactor = 0.5f;
    private int bottomInset;

    private HashMap<Integer, DateTimeFormatter> _clockModes = new HashMap<>(4);
    private Integer _clockFormatterIndex = -1;
    private DateTimeFormatter _clockFormatter = null;

    public SearchBar(@NonNull Context context) {
        super(context);
        init();
    }

    public SearchBar(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public SearchBar(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public void setCallback(CallBack callback) {
        _callback = callback;
    }

    public boolean collapse() {
        if (!_expanded) {
            return false;
        }
        _searchButton.callOnClick();
        return !_expanded;
    }

    private void init() {
        int dp1 = Tool.dp2px(1);
        int iconMarginOutside = dp1 * 16;
        int iconMarginTop = dp1 * 14;
        int searchTextMarginTop = dp1 * 4;
        int iconSize = dp1 * 24;
        int iconPadding = dp1 * 6;

        // These have to match the Preferences Array, but without item 0 as that is a custom option which can be changed:
        //   <item>@string/custom</item>
        //   <item>February 17\nSaturday, 2018</item>
        //   <item>February 17\n15:48</item>
        //   <item>February 17, 2018\n15:48</item>
        //   <item>15:48\nFebruary 17, 2018</item>
        _clockModes.put(1, DateTimeFormatter.ofPattern("MMMM dd\nEEEE, yyyy", Locale.getDefault()));
        _clockModes.put(2, DateTimeFormatter.ofPattern("MMMM dd\nHH:mm", Locale.getDefault()));
        _clockModes.put(3, DateTimeFormatter.ofPattern("MMMM dd, yyyy\nHH:mm", Locale.getDefault()));
        _clockModes.put(4, DateTimeFormatter.ofPattern("HH:mm\nMMMM dd, yyyy", Locale.getDefault()));

        _searchClock = (TextView) LayoutInflater.from(getContext()).inflate(R.layout.view_search_clock, this, false);
        _searchClock.setTextSize(TypedValue.COMPLEX_UNIT_DIP, _searchClockTextSize);
        LayoutParams clockParams = new LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        clockParams.setMargins(iconMarginOutside, dp1 * 4, 0, dp1 * 4);
        clockParams.gravity = Gravity.START;

        _switchButton = new AppCompatImageView(getContext());
        _switchButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                Setup.appSettings().setSearchUseGrid(!Setup.appSettings().getSearchUseGrid());
                updateSwitchIcon();
                updateRecyclerViewLayoutManager();
            }
        });
        _switchButton.setVisibility(View.GONE);
        _switchButton.setPadding(0, iconPadding, 0, iconPadding);
        updateSwitchIcon();

        LayoutParams switchButtonParams = new LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        switchButtonParams.setMargins(iconMarginOutside / 2, 0, 0, 0);
        switchButtonParams.gravity = Gravity.START | Gravity.CENTER_VERTICAL;

        if (isInEditMode()) return;

        _icon = new CircleDrawable(getContext(), getResources().getDrawable(R.drawable.ic_search), Color.WHITE, Color.BLACK, 100);
        _searchButton = new AppCompatImageView(getContext());
        _searchButton.setImageDrawable(_icon);
        _searchButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                if (_expanded && _searchInput.getText().length() > 0) {
                    _searchInput.getText().clear();
                    return;
                }
                _expanded = !_expanded;
                if (_expanded) {
                    expandInternal();
                } else {
                    collapseInternal();
                }
            }
        });

        LayoutParams buttonParams = new LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        buttonParams.setMargins(0, iconMarginTop, iconMarginOutside, 0);
        buttonParams.gravity = Gravity.END;

        _searchCardContainer = new CardView(getContext());
        _searchCardContainer.setCardBackgroundColor(Color.TRANSPARENT);
        _searchCardContainer.setVisibility(View.GONE);
        _searchCardContainer.setRadius(0);
        _searchCardContainer.setCardElevation(0);
        _searchCardContainer.setContentPadding(dp1 * 4, dp1 * 4, dp1 * 4, dp1 * 4);

        _searchInput = new AppCompatEditText(getContext());
        _searchInput.setBackground(null);
        _searchInput.setHint(R.string.search_hint);
        _searchInput.setHintTextColor(Color.WHITE);
        _searchInput.setTextColor(Color.WHITE);
        _searchInput.setSingleLine();
        _searchInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                _adapter.filter(s);
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });

        _searchInput.setOnKeyListener(new OnKeyListener() {
            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if ((event != null) && (event.getAction() == KeyEvent.ACTION_DOWN) && (keyCode == KeyEvent.KEYCODE_ENTER)) {
                    _callback.onInternetSearch(_searchInput.getText().toString());
                    _searchInput.getText().clear();
                    return true;
                }
                return false;
            }
        });
        LayoutParams inputCardParams = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        inputCardParams.setMargins(0, searchTextMarginTop, 0, 0);

        LayoutParams inputParams = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        inputParams.setMargins(iconMarginOutside + iconSize, 0, 0, 0);

        _searchCardContainer.addView(_switchButton, switchButtonParams);
        _searchCardContainer.addView(_searchInput, inputParams);

        initRecyclerView();

        Setup.appLoader().addUpdateListener(new AppUpdateListener() {
            @Override
            public boolean onAppUpdated(List<App> apps) {
                _adapter.clear();
                if (Setup.appSettings().getSearchBarShouldShowHiddenApps()) {
                    apps = Setup.appLoader().getAllApps(getContext(), true);
                }
                List<IconLabelItem> items = new ArrayList<>();
                for (int i = 0; i < apps.size(); i++) {
                    final App app = apps.get(i);
                    items.add(new IconLabelItem(app.getIcon(), app.getLabel())
                            .withIconSize(50)
                            .withTextColor(Color.WHITE)
                            .withIsAppLauncher(true)
                            .withIconPadding(8)
                            .withOnClickAnimate(false)
                            .withTextGravity(Setup.appSettings().getSearchUseGrid() ? Gravity.CENTER : Gravity.CENTER_VERTICAL)
                            .withIconGravity(Setup.appSettings().getSearchUseGrid() ? Gravity.TOP : Gravity.START)
                            .withOnClickListener(new OnClickListener() {
                                @Override
                                public void onClick(View v) {
                                    Tool.startApp(v.getContext(), app, null);
                                }
                            })
                            .withOnLongClickListener(DragHandler.getLongClick(Item.newAppItem(app), DragAction.Action.SEARCH, null)));
                }
                _adapter.set(items);

                return false;
            }
        });
        _adapter.getItemFilter().withFilterPredicate(new IItemAdapter.Predicate<IconLabelItem>() {
            @Override
            public boolean filter(IconLabelItem item, CharSequence constraint) {
                if (constraint.length() == 0) {
                    return true;
                }

                String s = constraint.toString().toLowerCase();
                s = Normalizer.normalize(s, Form.NFD).replaceAll("\\p{InCombiningDiacriticalMarks}+", "");
                String itemLabel = item._label.toLowerCase();
                itemLabel = Normalizer.normalize(itemLabel, Form.NFD).replaceAll("\\p{InCombiningDiacriticalMarks}+", "");

                if (Setup.appSettings().getSearchBarStartsWith()) {
                    if (itemLabel.startsWith(s)) {
                        return true;
                    }
                } else {
                    if (itemLabel.contains(s)) {
                        return true;
                    }
                }

                return false;
            }
        });

        final LayoutParams recyclerParams = new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);

        addView(_searchClock, clockParams);
        addView(_searchRecycler, recyclerParams);
        addView(_searchCardContainer, inputCardParams);
        addView(_searchButton, buttonParams);

        _searchInput.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                _searchInput.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                int marginTop = Tool.dp2px(60);
                recyclerParams.setMargins(0, marginTop, 0, 0);
                _searchRecycler.setLayoutParams(recyclerParams);
                _searchRecycler.setPadding(0, 0, 0, (int) (bottomInset * 1.5));
            }
        });
    }

    private void collapseInternal() {
        if (_callback != null) {
            _callback.onCollapse();
        }

        _icon.setIcon(getResources().getDrawable(R.drawable.ic_search));

        Tool.visibleViews(ANIM_TIME, _searchClock);
        Tool.goneViews(ANIM_TIME, _searchCardContainer, _searchRecycler, _switchButton);

        _searchInput.getText().clear();
    }

    private void expandInternal() {
        if (_callback != null) {
            _callback.onExpand();
        }

        _icon.setIcon(getResources().getDrawable(R.drawable.ic_clear));

        Tool.visibleViews(ANIM_TIME, _searchCardContainer, _searchRecycler, _switchButton);
        Tool.goneViews(ANIM_TIME, _searchClock);
    }

    private void updateSwitchIcon() {
        _switchButton.setImageResource(Setup.appSettings().getSearchUseGrid() ? R.drawable.ic_view_grid_white : R.drawable.ic_view_list_white);
    }

    private void updateRecyclerViewLayoutManager() {
        int gridSize = Setup.appSettings().getSearchUseGrid() ? Setup.appSettings().getDrawerColumnCount() : 1;
        if (gridSize == 1) {
            _searchRecycler.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
            updateList(Gravity.START, Gravity.CENTER_VERTICAL);
        } else {
            _searchRecycler.setLayoutManager(new GridLayoutManager(getContext(), gridSize, GridLayoutManager.VERTICAL, false));
            updateList(Gravity.TOP, Gravity.CENTER);
        }
        _searchRecycler.getLayoutManager().setAutoMeasureEnabled(false);
    }

    private void updateList(int iconGravity, int textGravity) {
        List<IconLabelItem> apps = _adapter.getAdapterItems();
        for (IconLabelItem app : apps) {
            app.setIconGravity(iconGravity);
            app.setTextGravity(textGravity);
        }
    }

    protected void initRecyclerView() {
        _searchRecycler = new RecyclerView(getContext());
        _searchRecycler.setItemAnimator(null);
        _searchRecycler.setVisibility(View.GONE);
        _searchRecycler.setAdapter(_adapter);
        _searchRecycler.setClipToPadding(false);
        _searchRecycler.setHasFixedSize(true);
        updateRecyclerViewLayoutManager();
    }

    public AppCompatImageView getSearchButton() {
        return _searchButton;
    }

    public void updateClock() {
        AppSettings appSettings = AppSettings.get();
        if (_searchClock != null) {
            _searchClock.setTextColor(appSettings.getDesktopDateTextColor());
        }

        ZonedDateTime now = ZonedDateTime.now();
        _clockFormatter = getSearchBarClockFormat(Setup.appSettings().getDesktopDateMode());

        String text = now.format(_clockFormatter);
        String[] lines = text.split("\n");
        if (lines.length < 2) {
            _searchClock.setText(lines[0]);
        } else {
            Spannable span = new SpannableString(text);
            span.setSpan(new RelativeSizeSpan(_searchClockSubTextFactor), lines[0].length() + 1, lines[0].length() + 1 + lines[1].length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            _searchClock.setText(span);
        }
    }

    @Override
    public WindowInsets onApplyWindowInsets(WindowInsets insets) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH) {
            bottomInset = insets.getSystemWindowInsetBottom();
            setPadding(0, insets.getSystemWindowInsetTop(), 0, 0);
            return insets;
        }
        return insets;
    }

    public DateTimeFormatter getSearchBarClockFormat(Integer id) {
        if (_clockFormatterIndex != id && id > 0) {
            if (_clockModes.containsKey(id)) {
                return _clockModes.get(id);
            }
        }

        return Setup.appSettings().getUserDateFormat();
    }

    public interface CallBack {
        void onInternetSearch(String string);

        void onExpand();

        void onCollapse();
    }
}

File: app/src/main/java/com/benny/openlauncher/widget/SearchBar.kt
package com.benny.openlauncher.widget

import android.content.Context
import android.graphics.Color
import android.os.Build
import android.support.annotation.AttrRes
import android.support.v7.widget.*
import android.text.Editable
import android.text.Spannable
import android.text.SpannableString
import android.text.TextWatcher
import android.text.style.RelativeSizeSpan
import android.util.AttributeSet
import android.util.TypedValue
import android.view.*
import android.view.View.OnClickListener
import android.widget.FrameLayout
import android.widget.TextView
import com.benny.openlauncher.R
import com.benny.openlauncher.drawable.LauncherCircleDrawable
import com.benny.openlauncher.interfaces.AbstractApp
import com.benny.openlauncher.interfaces.AppUpdateListener
import com.benny.openlauncher.interfaces.SettingsManager
import com.benny.openlauncher.manager.Setup
import com.benny.openlauncher.model.IconLabelItem
import com.benny.openlauncher.model.Item
import com.benny.openlauncher.util.DragAction
import com.benny.openlauncher.util.Tool
import com.mikepenz.fastadapter.IItemAdapter
import com.mikepenz.fastadapter.commons.adapters.FastItemAdapter
import java.text.SimpleDateFormat
import java.util.*

class SearchBar : FrameLayout {

    val searchClock: TextView
    val switchButton: AppCompatImageView
    val searchButton: AppCompatImageView
    val searchInput: AppCompatEditText
    val searchRecycler: RecyclerView

    private val icon: LauncherCircleDrawable
    private val icon2: LauncherCircleDrawable

    private val searchCardContainer: FrameLayout
    private val adapter = FastItemAdapter<IconLabelItem>()
    private var callback: CallBack? = null
    private var expanded: Boolean = false
    private var searchInternetEnabled = true
    private var mode = Mode.DateAll
    private var searchClockTextSize = 28
    private var searchClockSubTextFactor = 0.5f

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, @AttrRes defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    fun setSearchInternetEnabled(enabled: Boolean): SearchBar {
        searchInternetEnabled = enabled
        return this
    }

    fun setSearchClockTextSize(size: Int): SearchBar {
        searchClockTextSize = size
        if (searchClock != null) {
            searchClock!!.setTextSize(TypedValue.COMPLEX_UNIT_DIP, searchClockTextSize.toFloat())
        }
        return this
    }

    fun setSearchClockSubTextFactor(factor: Float): SearchBar {
        searchClockSubTextFactor = factor
        return this
    }

    fun setMode(mode: Mode): SearchBar {
        this.mode = mode
        return this
    }

    fun setCallback(callback: CallBack) {
        this.callback = callback
    }

    fun collapse(): Boolean {
        if (!expanded) {
            return false
        }
        searchButton.callOnClick()
        return !expanded
    }

    init {
        val dp1 = Tool.dp2px(1, context)
        val iconMarginOutside = dp1 * 16
        val iconMarginTop = dp1 * 13
        val searchTextHorizontalMargin = dp1 * 8
        val searchTextMarginTop = dp1 * 4
        val iconSize = dp1 * 24
        val iconPadding = dp1 * 6 // LauncherCircleDrawable uses 6dp as well!!

        searchClock = LayoutInflater.from(context).inflate(R.layout.view_search_clock, this, false) as TextView
        searchClock!!.setTextSize(TypedValue.COMPLEX_UNIT_DIP, searchClockTextSize.toFloat())
        val clockParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        clockParams.setMargins(iconMarginOutside, 0, 0, 0)
        clockParams.gravity = Gravity.START

        // && Setup.appSettings().isSearchGridListSwitchEnabled()

        icon2 = LauncherCircleDrawable(context, resources.getDrawable(if (Setup.appSettings().isSearchUseGrid) R.drawable.ic_apps_white_48dp else R.drawable.ic_view_list_white_24dp), Color.BLACK)
        switchButton = AppCompatImageView(context)
        switchButton.setImageDrawable(icon2)
        switchButton.setOnClickListener {
            Setup.appSettings().isSearchUseGrid = !Setup.appSettings().isSearchUseGrid
            updateSwitchIcon()
            updateRecyclerViewLayoutManager()
        }

        val switchButtonParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        switchButtonParams.setMargins(0, 0, iconSize + iconMarginOutside * 2, 0)
        switchButtonParams.gravity = Gravity.END or Gravity.CENTER_VERTICAL

        searchButton = AppCompatImageView(context)
        searchInput = AppCompatEditText(context)

        icon = LauncherCircleDrawable(context, resources.getDrawable(R.drawable.ic_search_light_24dp), Color.BLACK)
        searchButton.setImageDrawable(icon)
        searchButton.setOnClickListener(OnClickListener {
            if (expanded && searchInput.text.isNotEmpty()) {
                searchInput.text.clear()
                return@OnClickListener
            }
            expanded = !expanded
            if (expanded) {
                expandInternal()
            } else {
                collapseInternal()
            }
        })
        val buttonParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        buttonParams.setMargins(0, iconMarginTop, iconMarginOutside, 0)
        buttonParams.gravity = Gravity.END

        searchCardContainer = FrameLayout(context)
        //searchCardContainer.setOverlayColor(resources.getColor(R.color.colorAccent))
        searchCardContainer.visibility = View.GONE
        searchCardContainer.setPadding(dp1 * 4, dp1 * 4, dp1 * 4, dp1 * 4)

//        val root = CoreHome.launcher!!.window.decorView.findViewById<ViewGroup>(android.R.id.content)
//        searchCardContainer.setupWith(root)
//                .windowBackground(root.background)
//                .blurAlgorithm(RenderScriptBlur(context))
//                .blurRadius(20f)

        searchInput.background = null
        searchInput.setHint(R.string.search_hint)
        searchInput.setHintTextColor(Color.WHITE)
        searchInput.setTextColor(Color.WHITE)
        searchInput.setSingleLine()
        searchInput.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (searchInput.text.isEmpty()) {
                    clearFilter()
                } else
                    adapter.filter(s)
            }

            override fun afterTextChanged(s: Editable) {

            }
        })
        searchInput.setOnKeyListener(object : View.OnKeyListener {
            override fun onKey(v: View?, keyCode: Int, event: KeyEvent?): Boolean {
                if ((event != null) && (event.action == KeyEvent.ACTION_DOWN) && (keyCode == KeyEvent.KEYCODE_ENTER)) {
                    callback!!.onInternetSearch(searchInput.text.toString())
                    searchInput.text.clear()
                    return true
                }
                return false
            }
        })
        val inputCardParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        //(switchButton != null ? iconMarginOutside + iconSize : 0) + searchTextHorizontalMargin
        //iconMarginOutside + iconSize + searchTextHorizontalMargin
        inputCardParams.setMargins(0, searchTextMarginTop, 0, 0)

        val inputParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        inputParams.setMargins(iconMarginOutside / 2, 0, 0, 0)

        searchCardContainer.addView(searchInput, inputParams)
        searchCardContainer.addView(switchButton, switchButtonParams)

        searchRecycler = RecyclerView(context)
        searchRecycler.overScrollMode = View.OVER_SCROLL_NEVER
        searchRecycler.itemAnimator = null
        searchRecycler.visibility = View.GONE
        searchRecycler.adapter = adapter
        searchRecycler.clipToPadding = false
        searchRecycler.setHasFixedSize(true)

        Setup.appLoader().addUpdateListener(AppUpdateListener { apps ->
            var apps = apps
            adapter.clear()
            if (Setup.appSettings().searchBarShouldShowHiddenApps) {
                apps = Setup.appLoader().getAllApps(context, true)
            }
            val items = ArrayList<IconLabelItem>()
            if (searchInternetEnabled) {
                items.add(IconLabelItem(context, R.drawable.ic_search_light_24dp, R.string.search_online)
                        .withIconGravity(if (Setup.appSettings().isSearchUseGrid) Gravity.TOP else Gravity.LEFT)
                        .withTextGravity(if (Setup.appSettings().isSearchUseGrid) Gravity.CENTER_HORIZONTAL else Gravity.RIGHT)
                        .withOnClickListener {
                            callback!!.onInternetSearch(searchInput.text.toString())
                            searchInput.text.clear()
                        }
                        .withTextColor(Color.WHITE)
                        .withDrawablePadding(context, 19)
                        .withBold(true)
                        .withMatchParent(true))
            }
            for (i in apps.indices) {
                val app = apps[i]
                items.add(IconLabelItem(context, app.iconProvider, app.label, 36)
                        .withIconGravity(if (Setup.appSettings().isSearchUseGrid) Gravity.TOP else Gravity.LEFT)
                        .withTextGravity(if (Setup.appSettings().isSearchUseGrid) Gravity.CENTER_HORIZONTAL else Gravity.CENTER_VERTICAL)
                        .withOnClickListener { v -> startApp(v.context, app, v) }
                        .withOnTouchListener(AppItemView.Builder.getOnTouchGetPosition(null, null))
                        .withOnLongClickListener(true, AppItemView.Builder.getLongClickDragAppListener(Item.newAppItem(app), DragAction.Action.SEARCH_RESULT, object : AppItemView.LongPressCallBack {
                            override fun readyForDrag(view: View): Boolean {
                                expanded = !expanded
                                collapseInternal()
                                return true
                            }

                            override fun afterDrag(view: View) {}
                        }))
                        .withTextColor(Color.WHITE)
                        .withMatchParent(true)
                        .withDrawablePadding(context, 8)
                        .withMaxTextLines(Setup.appSettings().searchLabelLines))
            }
            adapter.set(items)
            adapter.withOnLongClickListener(com.mikepenz.fastadapter.listeners.OnLongClickListener { v, adapter, item, position ->
                if (!searchInternetEnabled || position != 0) {
                    item.onLongClickListener.onLongClick(v)
                    return@OnLongClickListener true
                }
                false
            })
            clearFilter()
            false
        })
        adapter.itemFilter.withFilterPredicate(IItemAdapter.Predicate { item, constraint ->
            updateItemGravity(if (item.label == context.getString(R.string.search_online)) 0 else 1, item)
            if (item.label == context.getString(R.string.search_online))
                return@Predicate true
            val s = constraint!!.toString()
            if (s.isBlank())
                return@Predicate true
            else
                return@Predicate item.label.toLowerCase().contains(s.toLowerCase())
        })
        updateRecyclerViewLayoutManager()

        val recyclerParams = FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)

        addView(searchClock, clockParams)
        addView(searchRecycler, recyclerParams)
        addView(searchCardContainer, inputCardParams)
        addView(searchButton, buttonParams)

        searchInput.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                searchInput.viewTreeObserver.removeOnGlobalLayoutListener(this)
                val marginTop = Tool.dp2px(56, context) + searchInput.height
                val marginBottom = Desktop.bottomInset
                recyclerParams.setMargins(0, marginTop, 0, marginBottom)
                recyclerParams.height = (parent as View).height - marginTop - marginBottom / 2
                searchRecycler.layoutParams = recyclerParams
                searchRecycler.setPadding(0, 0, 0, (marginBottom * 1.5f).toInt())
            }
        })
    }

    private fun clearFilter() {
        adapter.filter(" ")
    }

    private fun collapseInternal() {
        if (callback != null) {
            callback!!.onCollapse()
        }
        icon!!.setIcon(resources.getDrawable(R.drawable.ic_search_light_24dp))
        Tool.visibleViews(ANIM_TIME, searchClock!!)
        Tool.goneViews(ANIM_TIME, searchCardContainer, searchRecycler, switchButton)
        searchInput.text.clear()
    }

    private fun expandInternal() {
        if (callback != null) {
            callback!!.onExpand()
        }
        if (Setup.appSettings().isResetSearchBarOnOpen) {
            val lm = searchRecycler.layoutManager
            if (lm is LinearLayoutManager) {
                (searchRecycler.layoutManager as LinearLayoutManager).scrollToPositionWithOffset(0, 0)
            } else if (lm is GridLayoutManager) {
                (searchRecycler.layoutManager as GridLayoutManager).scrollToPositionWithOffset(0, 0)
            }
        }
        icon!!.setIcon(resources.getDrawable(R.drawable.ic_clear_white_24dp))
        Tool.visibleViews(ANIM_TIME, searchCardContainer, searchRecycler, switchButton)
        Tool.goneViews(ANIM_TIME, searchClock!!)
    }

    private fun updateSwitchIcon() {
        icon2.setIcon(resources.getDrawable(if (Setup.appSettings().isSearchUseGrid) R.drawable.ic_apps_white_48dp else R.drawable.ic_view_list_white_24dp))
    }

    private fun updateRecyclerViewLayoutManager() {
        for ((i, item) in adapter.adapterItems.withIndex()) {
            updateItemGravity(i, item)
        }
        adapter.notifyAdapterDataSetChanged()

        val gridSize = if (Setup.appSettings().isSearchUseGrid) Setup.appSettings().searchGridSize else 1
        if (gridSize == 1) {
            searchRecycler.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        } else {
            searchRecycler.layoutManager = GridLayoutManager(context, gridSize, GridLayoutManager.VERTICAL, false)
        }
        searchRecycler.layoutManager.isAutoMeasureEnabled = false
    }

    private fun updateItemGravity(position: Int, item: IconLabelItem) {
        if (position == 0) {
            item.withTextGravity(if (Setup.appSettings().isSearchUseGrid) Gravity.CENTER_HORIZONTAL else Gravity.RIGHT)
        } else {
            item.withTextGravity(if (Setup.appSettings().isSearchUseGrid) Gravity.CENTER_HORIZONTAL else Gravity.CENTER_VERTICAL)
        }
        item.withIconGravity(if (Setup.appSettings().isSearchUseGrid) Gravity.TOP else Gravity.LEFT)
    }

    protected fun startApp(context: Context, app: AbstractApp, view: View) {
        Tool.startApp(context, app, view)
    }

    fun updateClock() {
        val settingsManager: SettingsManager =  Setup.appSettings()
        if (!settingsManager.isSearchBarTimeEnabled) {
            searchClock!!.text = ""
            return
        }
        searchClock!!.setTextColor(settingsManager.desktopDateTextColor)
        val calendar = Calendar.getInstance(Locale.getDefault())


        var sdf: SimpleDateFormat? = mode.sdf
        val mode = settingsManager.desktopDateMode
        if (mode >= 0 && mode < Mode.count()){
            sdf = Mode.getById(mode).sdf
            if (mode == 0){
                sdf = settingsManager.userDateFormat
            }
        }

        if (sdf == null) {
            sdf = Setup.appSettings().userDateFormat
        }
        val text = sdf!!.format(calendar.time)
        val lines = text.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val span = SpannableString(text)
        span.setSpan(RelativeSizeSpan(searchClockSubTextFactor), lines[0].length + 1, lines[0].length + 1 + lines[1].length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
        searchClock!!.text = span
    }

    override fun onApplyWindowInsets(insets: WindowInsets): WindowInsets {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT_WATCH) {
            setPadding(paddingLeft, insets.systemWindowInsetTop + Tool.dp2px(10, context), paddingRight, paddingBottom)
        }
        return insets
    }

    enum class Mode private constructor(id: Int, internal var sdf: SimpleDateFormat?) {
        DateAll(1, SimpleDateFormat("MMMM dd'\n'EEEE',' yyyy", Locale.getDefault())),
        DateNoYearAndTime(2, SimpleDateFormat("MMMM dd'\n'HH':'mm", Locale.getDefault())),
        DateAllAndTime(3, SimpleDateFormat("MMMM dd',' yyyy'\n'HH':'mm", Locale.getDefault())),
        TimeAndDateAll(4, SimpleDateFormat("HH':'mm'\n'MMMM dd',' yyyy", Locale.getDefault())),
        Custom(0, null);

        var id: Int = 0
            internal set

        init {
            this.id = id
        }

        companion object {

            fun getById(id: Int): Mode {
                for (i in 0 until values().size) {
                    if (values()[i].id == id)
                        return values()[i]
                }
                throw RuntimeException("ID not found!")
            }

            fun getByIndex(index: Int): Mode {
                return values()[index]
            }

            fun getIndex(id: Int): Int {
                for (i in 0 until values().size) {
                    if (values()[i].id == id) {
                        return i
                    }
                }
                throw RuntimeException("ID not found!")
            }
            fun count(): Int {
                return values().size;
            }
        }
    }

    interface CallBack {
        fun onInternetSearch(string: String)

        fun onExpand()

        fun onCollapse()
    }

    companion object {

        private val ANIM_TIME: Long = 200
    }
}
